File: buildforge/src/parser/parser.cpp
#include "parser.hpp"
#include <algorithm>

namespace buildforge {

Parser::Parser(std::vector<Token> tokens) : tokens_(std::move(tokens)) {}

const Token& Parser::peek() const {
    if (cursor_ >= tokens_.size()) return tokens_.back();
    return tokens_[cursor_];
}

const Token& Parser::previous() const {
    if (cursor_ == 0) return tokens_[0];
    return tokens_[cursor_ - 1];
}

Token Parser::advance() {
    if (cursor_ < tokens_.size()) {
        cursor_++;
    }
    return previous();
}

bool Parser::check(TokenType type) const {
    return peek().type == type;
}

bool Parser::match(TokenType type) {
    if (check(type)) {
        advance();
        return true;
    }
    return false;
}

bool Parser::consume(TokenType type, const std::string& error_message) {
    if (check(type)) {
        advance();
        return true;
    }
    diagnostics_.push_back({
        Diagnostic::Severity::Error,
        error_message,
        peek().location,
        "Expected " + error_message
    });
    return false;
}

bool Parser::is_label(const std::string& str) const {
    // Label format: //pkg:target or :target
    if (str.starts_with("//") && str.find(':') != std::string::npos) return true;
    if (str.starts_with(':')) return true;
    return false;
}

Result<std::vector<RuleCall>> Parser::parse() {
    std::vector<RuleCall> rules;
    while (!check(TokenType::Eof)) {
        auto rule = rule_call();
        if (rule) {
            rules.push_back(*rule);
        } else {
            // synchronize to next statement
            while (!check(TokenType::Eof) && !check(TokenType::Ident)) {
                advance();
            }
        }
    }
    if (!diagnostics_.empty()) {
        bool has_error = std::ranges::any_of(diagnostics_, [](const auto& d) {
            return d.severity == Diagnostic::Severity::Error;
        });
        if (has_error) {
            return std::unexpected(Error{
                StatusCode::InvalidArgument,
                "Parsing failed with " + std::to_string(diagnostics_.size()) + " diagnostics",
                "Parser"
            });
        }
    }
    return rules;
}

Result<RuleCall> Parser::rule_call() {
    if (!check(TokenType::Ident)) {
        diagnostics_.push_back({
            Diagnostic::Severity::Error,
            "Expected rule name (e.g. cc_binary, cc_library)",
            peek().location,
            "Make sure your target begins with an identifier"
        });
        return std::unexpected(Error{StatusCode::InvalidArgument, "Expected rule indentifier", "Parser"});
    }
    Token rule_tok = advance();
    SourceLocation rule_loc = rule_tok.location;

    if (!consume(TokenType::LPar, "Expected '(' after rule type name")) {
        return std::unexpected(Error{StatusCode::InvalidArgument, "Expected (", "Parser"});
    }

    std::unordered_map<std::string, Expr> args;
    std::string name;

    while (!check(TokenType::RPar) && !check(TokenType::Eof)) {
        auto arg = named_arg();
        if (!arg) {
            return std::unexpected(arg.error());
        }
        args[arg->name] = arg->value;
        if (arg->name == "name") {
            if (auto* str = std::get_if<StringLiteral>(&arg->value)) {
                name = str->value;
            }
        }
        if (!match(TokenType::Comma)) {
            if (!check(TokenType::RPar)) {
                diagnostics_.push_back({
                    Diagnostic::Severity::Error,
                    "Expected ',' or ')' inside named arguments",
                    peek().location,
                    "Add commas to separate target attributes"
                });
                return std::unexpected(Error{StatusCode::InvalidArgument, "Expected ',' or ')'", "Parser"});
            }
        }
    }

    if (!consume(TokenType::RPar, "Expected ')' to close rule arguments")) {
        return std::unexpected(Error{StatusCode::InvalidArgument, "Expected )", "Parser"});
    }

    RuleCall call{rule_tok.lexeme, name, args, rule_loc};
    return call;
}

Result<NamedArg> Parser::named_arg() {
    if (!check(TokenType::Ident)) {
        diagnostics_.push_back({
            Diagnostic::Severity::Error,
            "Expected argument key identifier (e.g. name, srcs, deps)",
            peek().location,
            "Specify an attribute name"
        });
        return std::unexpected(Error{StatusCode::InvalidArgument, "Expected argument key identifier", "Parser"});
    }
    Token key_tok = advance();
    if (!consume(TokenType::Equal, "Expected '=' after attribute key")) {
        return std::unexpected(Error{StatusCode::InvalidArgument, "Expected =", "Parser"});
    }

    auto val_res = expr();
    if (!val_res) {
        return std::unexpected(val_res.error());
    }

    return NamedArg{key_tok.lexeme, *val_res, key_tok.location};
}

Result<Expr> Parser::expr() {
    SourceLocation loc = peek().location;
    if (match(TokenType::String)) {
        std::string s = previous().lexeme;
        if (is_label(s)) {
            return LabelExpr{s, loc};
        } else {
            return StringLiteral{s, loc};
        }
    } else if (check(TokenType::LBracket)) {
        auto list_res = list_expr();
        if (!list_res) return std::unexpected(list_res.error());
        return *list_res;
    }

    diagnostics_.push_back({
        Diagnostic::Severity::Error,
        "Expected expression (string or list)",
        loc,
        "Provide a string literal or string array [src, ...]"
    });
    return std::unexpected(Error{StatusCode::InvalidArgument, "Expected expression", "Parser"});
}

Result<ListExpr> Parser::list_expr() {
    SourceLocation loc = peek().location;
    advance(); // Consume '['

    std::vector<std::variant<StringLiteral, LabelExpr>> elements;
    while (!check(TokenType::RBracket) && !check(TokenType::Eof)) {
        auto element_res = expr();
        if (!element_res) {
            return std::unexpected(element_res.error());
        }
        
        // Ensure element is not nested list
        if (auto* str = std::get_if<StringLiteral>(&*element_res)) {
            elements.push_back(*str);
        } else if (auto* lbl = std::get_if<LabelExpr>(&*element_res)) {
            elements.push_back(*lbl);
        } else {
            diagnostics_.push_back({
                Diagnostic::Severity::Error,
                "Lists cannot contain nested lists in simple BUILD format",
                peek().location,
                "Remove double brackets"
            });
            return std::unexpected(Error{StatusCode::InvalidArgument, "Nested list forbidden", "Parser"});
        }

        if (!match(TokenType::Comma)) {
            if (!check(TokenType::RBracket)) {
                diagnostics_.push_back({
                    Diagnostic::Severity::Error,
                    "Expected ',' or ']' after list item",
                    peek().location,
                    "Add a separating comma"
                });
                return std::unexpected(Error{StatusCode::InvalidArgument, "Expected ','", "Parser"});
            }
        }
    }

    if (!consume(TokenType::RBracket, "Expected ']' at the end of list")) {
        return std::unexpected(Error{StatusCode::InvalidArgument, "Expected ]", "Parser"});
    }

    return ListExpr{elements, loc};
}

} // namespace buildforge

File: src/lib/parser.ts
import { Token, TokenType, Diagnostic, RuleCall, Expr, SourceLocation, ListExpr, StringLiteral, LabelExpr } from '../types';

export class Lexer {
  private source: string;
  private filepath: string;
  private cursor = 0;
  private line = 1;
  private col = 1;

  constructor(source: string, filepath: string) {
    this.source = source;
    this.filepath = filepath;
  }

  private peek(): string {
    if (this.cursor >= this.source.length) return '\0';
    return this.source[this.cursor];
  }

  private get(): string {
    if (this.cursor >= this.source.length) return '\0';
    const c = this.source[this.cursor++];
    if (c === '\n') {
      this.line++;
      this.col = 1;
    } else {
      this.col++;
    }
    return c;
  }

  private skipWhitespaceAndComments() {
    while (this.cursor < this.source.length) {
      const c = this.peek();
      if (/\s/.test(c)) {
        this.get();
      } else if (c === '#') {
        while (this.cursor < this.source.length && this.peek() !== '\n') {
          this.get();
        }
      } else {
        break;
      }
    }
  }

  private lexString(): Token {
    const loc: SourceLocation = { path: this.filepath, line: this.line, col: this.col };
    const quote = this.get(); // ' or "
    let value = '';
    while (this.cursor < this.source.length && this.peek() !== quote) {
      if (this.peek() === '\\') {
        this.get(); // escape char
        if (this.cursor < this.source.length) {
          value += this.get();
        }
      } else {
        value += this.get();
      }
    }
    if (this.peek() === quote) {
      this.get(); // consume close quote
    }
    return { type: 'String', lexeme: value, location: loc };
  }

  private lexIdent(): Token {
    const loc: SourceLocation = { path: this.filepath, line: this.line, col: this.col };
    let value = '';
    while (this.cursor < this.source.length) {
      const c = this.peek();
      if (/[a-zA-Z0-9_]/.test(c)) {
        value += this.get();
      } else {
        break;
      }
    }
    return { type: 'Ident', lexeme: value, location: loc };
  }

  public tokenize(): Token[] {
    const tokens: Token[] = [];
    while (true) {
      this.skipWhitespaceAndComments();
      const c = this.peek();
      if (c === '\0') {
        tokens.push({ type: 'Eof', lexeme: '', location: { path: this.filepath, line: this.line, col: this.col } });
        break;
      }
      const loc: SourceLocation = { path: this.filepath, line: this.line, col: this.col };
      if (c === '(') {
        this.get();
        tokens.push({ type: 'LPar', lexeme: '(', location: loc });
      } else if (c === ')') {
        this.get();
        tokens.push({ type: 'RPar', lexeme: ')', location: loc });
      } else if (c === '[') {
        this.get();
        tokens.push({ type: 'LBracket', lexeme: '[', location: loc });
      } else if (c === ']') {
        this.get();
        tokens.push({ type: 'RBracket', lexeme: ']', location: loc });
      } else if (c === '=') {
        this.get();
        tokens.push({ type: 'Equal', lexeme: '=', location: loc });
      } else if (c === ',') {
        this.get();
        tokens.push({ type: 'Comma', lexeme: ',', location: loc });
      } else if (c === '"' || c === "'") {
        tokens.push(this.lexString());
      } else if (/[a-zA-Z_]/.test(c)) {
        tokens.push(this.lexIdent());
      } else {
        this.get();
        tokens.push({ type: 'Unknown', lexeme: c, location: loc });
      }
    }
    return tokens;
  }
}

export class Parser {
  private tokens: Token[];
  private cursor = 0;
  private diagnostics: Diagnostic[] = [];

  constructor(tokens: Token[]) {
    this.tokens = tokens;
  }

  private peek(): Token {
    if (this.cursor >= this.tokens.length) return this.tokens[this.tokens.length - 1];
    return this.tokens[this.cursor];
  }

  private previous(): Token {
    if (this.cursor === 0) return this.tokens[0];
    return this.tokens[this.cursor - 1];
  }

  private advance(): Token {
    if (this.cursor < this.tokens.length) {
      this.cursor++;
    }
    return this.previous();
  }

  private check(type: TokenType): boolean {
    return this.peek().type === type;
  }

  private match(type: TokenType): boolean {
    if (this.check(type)) {
      this.advance();
      return true;
    }
    return false;
  }

  private consume(type: TokenType, errorMessage: string): boolean {
    if (this.check(type)) {
      this.advance();
      return true;
    }
    this.diagnostics.push({
      severity: 'Error',
      message: errorMessage,
      location: this.peek().location,
      hint: `Expected token of type ${type}`
    });
    return false;
  }

  private isLabel(str: string): boolean {
    return (str.startsWith('//') && str.includes(':')) || str.startsWith(':');
  }

  public parse(): { rules: RuleCall[]; diagnostics: Diagnostic[] } {
    const rules: RuleCall[] = [];
    while (!this.check('Eof')) {
      try {
        const rule = this.ruleCall();
        if (rule) {
          rules.push(rule);
        } else {
          this.synchronize();
        }
      } catch (e) {
        this.synchronize();
      }
    }
    return { rules, diagnostics: this.diagnostics };
  }

  private synchronize() {
    this.advance();
    while (!this.check('Eof')) {
      if (this.previous().type === 'RPar') return;
      if (this.check('Ident')) return;
      this.advance();
    }
  }

  private ruleCall(): RuleCall | null {
    if (!this.check('Ident')) {
      this.diagnostics.push({
        severity: 'Error',
        message: 'Expected rule call name (e.g. cc_binary, cc_library)',
        location: this.peek().location,
        hint: 'Start your target definitions with standard rules'
      });
      return null;
    }

    const ruleTok = this.advance();
    const ruleType = ruleTok.lexeme;
    const ruleLoc = ruleTok.location;

    if (!this.consume('LPar', "Expected '(' after rule type name")) {
      return null;
    }

    const args: Record<string, Expr> = {};
    let name = '';

    while (!this.check('RPar') && !this.check('Eof')) {
      const arg = this.namedArg();
      if (!arg) return null;
      args[arg.name] = arg.value;
      if (arg.name === 'name' && arg.value.type === 'StringLiteral') {
        name = arg.value.value;
      }

      if (!this.match('Comma')) {
        if (!this.check('RPar')) {
          this.diagnostics.push({
            severity: 'Error',
            message: "Expected ',' or ')' inside named arguments list",
            location: this.peek().location,
            hint: 'Attributes within rules must be separated by commas'
          });
          return null;
        }
      }
    }

    if (!this.consume('RPar', "Expected ')' to close rule arguments")) {
      return null;
    }

    return {
      ruleType,
      name,
      args,
      location: ruleLoc
    };
  }

  private namedArg(): { name: string; value: Expr } | null {
    if (!this.check('Ident')) {
      this.diagnostics.push({
        severity: 'Error',
        message: 'Expected attribute parameter key identifier (e.g. name, srcs, deps)',
        location: this.peek().location,
        hint: 'Input key must be standard identifier'
      });
      return null;
    }

    const keyTok = this.advance();
    if (!this.consume('Equal', "Expected '=' after attribute key name")) {
      return null;
    }

    const val = this.expr();
    if (!val) return null;

    return { name: keyTok.lexeme, value: val };
  }

  private expr(): Expr | null {
    const loc = this.peek().location;
    if (this.match('String')) {
      const s = this.previous().lexeme;
      if (this.isLabel(s)) {
        return { type: 'LabelExpr', value: s, location: loc };
      } else {
        return { type: 'StringLiteral', value: s, location: loc };
      }
    } else if (this.check('LBracket')) {
      return this.listExpr();
    }

    this.diagnostics.push({
      severity: 'Error',
      message: 'Expected string literal or list bracket',
      location: loc,
      hint: 'Define lists with square brackets [] or strings with double quotes'
    });
    return null;
  }

  private listExpr(): ListExpr | null {
    const loc = this.peek().location;
    this.advance(); // consume LBracket

    const elements: (StringLiteral | LabelExpr)[] = [];
    while (!this.check('RBracket') && !this.check('Eof')) {
      const elementRes = this.expr();
      if (!elementRes) return null;

      if (elementRes.type === 'ListExpr') {
        this.diagnostics.push({
          severity: 'Error',
          message: 'Nested lists are not allowed in simple BUILD files',
          location: this.peek().location,
          hint: 'Keep array values inside flat string lists'
        });
        return null;
      }

      elements.push(elementRes);

      if (!this.match('Comma')) {
        if (!this.check('RBracket')) {
          this.diagnostics.push({
            severity: 'Error',
            message: "Expected ',' or ']' after list item",
            location: this.peek().location,
            hint: 'List elements must have comma separators'
          });
          return null;
        }
      }
    }

    if (!this.consume('RBracket', "Expected ']' at the end of list element")) {
      return null;
    }

    return {
      type: 'ListExpr',
      elements,
      location: loc
    };
  }
}
